packages/cli/src/cmds/compare/DiffLoader.ts

```diff
@@ -0,0 +1,125 @@
+import parseDiff from 'parse-diff';
+import { warn } from 'console';
+import assert from 'assert';
+
+import { executeCommand } from '../../lib/executeCommand';
+import { QueueObject, queue } from 'async';
+import { existsSync } from 'fs';
+
+export const MAX_DIFF_LENGTH = 1000 * 1000;
+
+export type Diff = {
+  command: string;
+  files: parseDiff.File[];
+};
+
+export class DiffLoaderQueue {
+  private diffQueue: QueueObject<string>;
+  private diffByRoots = new Map<string, Diff>();
+
+  constructor(public baseRevision: string, public headRevision: string) {
+    this.diffQueue = queue(this.executeDiffCommand.bind(this), 1);
+  }
+
+  async diff(sourcePathRoots: Set<string>): Promise<Diff | undefined> {
+    // Process only one diff at a time.
+
+    if (this.diffQueue.length()) await this.diffQueue.drain();
+
+    const roots = [...sourcePathRoots].sort().join(' ');
+
+    this.diffQueue.push(roots);
+    await this.diffQueue.drain();
+
+    return this.diffByRoots.get(roots);
```

Comment:
I feel this would be much cleaner and more convincingly correct using [p-limit](https://github.com/sindresorhus/p-limit). This queue stuff is fiddly.

Reply:
Thanks for the pointer. I did take a look at that - it's also a queue so, while I don't love the `queue` we are using, it's a known quantity to me at this point. At least the concept of a queue with a concurrency of 1 is easy to understand, even if the implementation isn't especially so.